model/src/main/kotlin/Vulnerability.kt

```diff
@@ -42,7 +40,7 @@ data class Vulnerability(
      * URL pointing to a website with information regarding the vulnerability.
      */
     @JsonInclude(JsonInclude.Include.NON_NULL)
-    val url: URL? = null
+    val url: String? = null
```

Comment:
Commit message:

> No specific features of the URL class are used, but it brings in some additional complexity (e.g. when doing equals checks).

I tend to argue that if a thing semantically is a URL, we should also call it a URL. This is for clarity and to avoid (accidental) misuse.

Anyway, I'm curious about the equals checks you mention. Could you provide a concrete example in the commit message to convince me?

> Other parts of the model also use strings to represent URLs, e.g. VcsInfo.

True. For the `VcsInfo` class the argument was that we need to be able to track CVS URLs that start with something like `:pserver:` or `:ext:`, which can't be represented by a Java `URL` nor the more relaxed (in terms of the allowed protocols) `URI` class.

Reply:
Reworded the commit message to be more concrete regarding equals checks. Removed the part with `VcsInfo` as it does not support the argument.

It is a valid point that URL-like things should be represented as URLs. Unfortunately, Java's `URL` class is really problematic in some areas. Especially the unexpected behaviour when testing for equality can cause trouble, e.g. unit tests that fail if they do not have an internet connection.

An alternative could be using the `URI` class instead. It does not suffer from the same problems.

Reply:
> An alternative could be using the `URI` class instead. It does not suffer from the same problems.

Then I'd have a tendency to use the `URI` class here. In fact, I would go as far as saying that we should *never* use the `URL` class in ORT, but always use the `URI` class for URLs (unless we can't e.g. because we need to use third-party APIs, of course), and only in cases like VcsInfo use `String` for the URL.

So: First choice `URI` class, second choice `String` class for URLs.

Opinions?

Reply:
Agree, then I change the type in both `Vulnerability` and `SecurityIssue` to `URI`.

Reply:
Thanks. @oheger-bosch could you then do another commit in this PR which changes *all* URL to URI in the ORT code base, for the same reason? I quickly checked, there should be < 20 occurrences.

Reply:
I had a look at the usages of the `URL` class in the codebase. In many places, the URL class is used to download data, which should be fine. (As the `URI` class does not offer such functionality, it cannot serve as a replacement here.)

Problematic are the cases where URLs occur as properties in classes. In this category, I only found one more hit in the data classes used by `ClearlyDefinedService`, which I have changed.